fix(media-stream): prevent camera light from staying on after call settings change [WPB-11786]

zskhan opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11786" title="WPB-11786" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-11786</a>  [Web] camera light stayed on after a 1:1 call
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->

Prevent getting multiple instances of media devices on AV Preferences screen

## Description
When a user visits Audio / Video Settings view, due to re rendering, it acquires multiple media streams.
Which are not destroyed when the component is completely removed from the DOM and re-rendered.

Re-rendering is happening because of `useKoSubscribableChildren`.
I moved all instances of `useKoSubscribableChildren` to the parent component(MainContent.tsx) and pass the required value to the child components (AVPreferences.tsx, CameraPreferences.tsx) and memoized them.
In this way now both of these components are rendered only once and hence only 1 media stream of Microphone and Camera is in use and properly destroyed on component destroy.

## Checklist

- [x] mentions the JIRA issue in the PR name (Ex. [WPB-XXXX])
- [x] PR has been self reviewed by the author;
- [ ] Hard-to-understand areas of the code have been commented;
- [ ] If it is a core feature, unit tests have been added;

zskhan committed:
fix(media-stream): prevent camera light from staying on after call settings change
prevent getting multiple instances of media devices on AV Preferences screen

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/wireapp/wire-webapp?pullRequest=19434) <br/>All committers have signed the CLA.

aweiss-dev commented on src/script/page/MainContent/MainContent.tsx:
```diff
@@ -187,6 +205,10 @@ const MainContent: FC<MainContentProps> = ({
                   callingRepository={repositories.calling}
                   mediaRepository={repositories.media}
                   propertiesRepository={repositories.properties}
+                  isMediaDevicesAreInitialized={isMediaDevicesAreInitialized}
```
areMediaDevicesInitialised is already a boolean, no need for 'is'

zskhan committed:
chore: review compliance

przemvs commented on src/script/page/MainContent/panels/preferences/avPreferences/CameraPreferences.tsx:
```diff
@@ -40,29 +39,25 @@ interface CameraPreferencesProps {
 }
 
-const CameraPreferences: React.FC<CameraPreferencesProps> = ({
+const CameraPreferencesComponent: React.FC<CameraPreferencesProps> = ({
```
While You updating component, please remove FC props :) And pass them next to props.

przemvs commented on src/script/page/MainContent/MainContent.tsx:
```diff
@@ -105,9 +107,25 @@ const MainContent: FC<MainContentProps> = ({
+    MediaDeviceType.AUDIO_OUTPUT,
+    MediaDeviceType.VIDEO_INPUT,
+  ]);
+  const {isMediaDevicesAreInitialized} = useInitializeMediaDevices(devicesHandler, mediaRepo.streamHandler);
```
Please be careful with this, it can throw some of errors :)

zskhan replied:
Can you please guide me what should I care for?

zskhan committed:
fix(camera-preferences): update component signature

e-maad commented on src/script/page/MainContent/panels/preferences/avPreferences/CameraPreferences.tsx:
```diff
@@ -80,16 +75,18 @@ const CameraPreferences: React.FC<CameraPreferencesProps> = ({
       }
     } catch (error) {
-      logger.warn(`Requesting MediaStream for type "${MediaType.VIDEO}" failed: ${error.message}`, error);
+      if (error instanceof Error) {
```
Do we need this check?

zskhan replied:
Not needed but my IDE was giving error so I added this check as it was also there in other places.

EnricoSchw commented:
@zskhan I see what you're trying to do.

I think, because you've moved access to the device handler to a higher component, the app will now aske for video and audio permissions immediately when the the user logging in.

However, we only want to ask for permissions when we do the first call. React is not my strongest point, do you see a way around that?

EnricoSchw reviewed (changes requested):
Let us discuss first my question before you merge, please.

zskhan committed:
fix(main-content): move the useInitializeMediaDevice to AVPreferences.tsx
add useCallback to the method of hook
return true if media devices are initialized.

zskhan commented:
> @zskhan I see what you're trying to do.
> 
> I think, because you've moved access to the device handler to a higher component, the app will now aske for video and audio permissions immediately when the the user logging in.
> 
> However, we only want to ask for permissions when we do the first call. React is not my strongest point, do you see a way around that?

@EnricoSchw As we discussed this thing yesterday, I fixed the issue.
Whenever you have time please take a look.
I verified the ongoing call scenario also.

EnricoSchw commented:
> @EnricoSchw As we discussed this thing yesterday, I fixed the issue.
> Whenever you have time please take a look.
> I verified the ongoing call scenario also.

@zskhan Good work it works 👍
